### PR TITLE
[Bugfix] Fix CORS SSRF security issue

### PR DESCRIPTION
**Pull request by fred-bf**
Fix CORS SSRF security issue reported by: https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/issues/4283

- [x] Migrate Upstash API
- [x] Migrate WebDAV API
- [x] Remove request to `CORS` api

### CODE REVIEW

**Comment by fred-bf** [comment]
> by the way, for the webdav its difficult to filter

My current idea is to filter requests by folder and fileName

**Comment by fred-bf** [comment]
the folder name cannot be changed by user right now

**Comment by H0llyW00dzZ** [comment]
> > by the way, for the webdav its difficult to filter
> 
> My current idea is to filter requests by folder and fileName

There's another approach beyond using folder/fileName.

This approach involves:

Verifying methods for requests. Then, in this [`one`](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/blob/main/app/api/cors/%5B...path%5D/route.ts), apply it exclusively only for syncing WebDAV, Upstah, or others syncing that for implementing in future.


Ref:

- https://easy-software.com/en/glossary/webdav-protocol/

**Comment by fred-bf** [comment]
> > > by the way, for the webdav its difficult to filter
> > 
> > 
> > My current idea is to filter requests by folder and fileName
> 
> There's another approach beyond using folder/fileName.
> 
> This approach involves:
> 
> Verifying methods for requests. Then, in this [`one`](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/blob/main/app/api/cors/%5B...path%5D/route.ts), apply it exclusively only for syncing WebDAV, Upstah, or others syncing that for implementing in future.
> 
> Ref:
> 
> * https://easy-software.com/en/glossary/webdav-protocol/


Yes, it is indeed easier to maintain from a development perspective.  I'm worried that the loopholes here may be overlooked in the future if we put them all together. It would be more intuitive for each specific interface to maintain the forwarding capability separately.

**Comment by H0llyW00dzZ** [comment]
> > > > by the way, for the webdav its difficult to filter
> > > 
> > > 
> > > My current idea is to filter requests by folder and fileName
> > 
> > 
> > There's another approach beyond using folder/fileName.
> > This approach involves:
> > Verifying methods for requests. Then, in this [`one`](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/blob/main/app/api/cors/%5B...path%5D/route.ts), apply it exclusively only for syncing WebDAV, Upstah, or others syncing that for implementing in future.
> > Ref:
> > 
> > * https://easy-software.com/en/glossary/webdav-protocol/
> 
> Yes, it is indeed easier to maintain from a development perspective. I'm worried that the loopholes here may be overlooked in the future if we put them all together. It would be more intuitive for each specific interface to maintain the forwarding capability separately.

No need to worry, it's possible to modularize it separately, like this example:

image

That method switch golang

File: app/api/upstash/[action]/[...key]/route.ts [hunk 1]
```diff
@@ -2,13 +2,42 @@ import { NextRequest, NextResponse } from "next/server";
...
 ) {
+  const requestUrl = new URL(req.url);
+  const endpoint = requestUrl.searchParams.get("endpoint");
+
   if (req.method === "OPTIONS") {
     return NextResponse.json({ body: "OK" }, { status: 200 });
   }
+  const [action, ...key] = params.key;
+  // only allow to request to *.upstash.io
+  if (!endpoint || !endpoint.endsWith("upstash.io")) {
```
Comment (jokester) [inline app/api/upstash/[action]/[...key]/route.ts:15]: is string `endsWith` enough here? It seems still possible to craft a endpoint like `https://arbitrary.url.i.want/path?dummy=upstash.io` and bypass.

What about 

```diff
try {
  if ( ! (new URL(endpoint).hostname.endswith('.upstash.io')   )  {
   throw 403
 }
} catch (e) {
  throw 400
}

Comment (fred-bf) [reply app/api/upstash/[action]/[...key]/route.ts]: Thanks for reviewing the code. Just updated with the new matching method.

**Review by jokester — APPROVED** [review]
lgtm!!!

This should make it much harder to do bad things with `upstash` / `webdav` APIs.

for `app/api/cors/[...path]/route.ts` I suppose some fix is still needed.

**Comment by Yidadaa** [comment]
Let's turn it into a RPC, here is a proposal:

```ts
type WebDavConfig = {
  type: 'webdav',
  endpoint: string,
  username: string,
  password: string
}

type UpstashConfig = {
  type: 'upstash',
  endpoint: string,
  username: STORAGE_KEY,
  apiKey: string
}

type SyncAction = {
  type: 'check'
} | {
  type: 'get',
  key: string,
} | {
  type: 'set',
  key: string,
  value: string
}

type SyncPayload = {
  config: WebDavConfig | UpstashConfig,
  action: SyncAction
}

// client side
function createUpstashClient() {
  return async check() {
    return await call('/api/sync', { /* webdav config */ })
  },
  // get(key), set(key, value)
}

// server side, /api/sync/route.ts
async function handle(req) {
  const payload = JSON.parse(await req.json()) as SyncPayload;

  // dispatch payload to server side check/get/set actions
}
```

The main idea is to move all the sync actions that were previously done on the client side to the server side. The downside is that the desktop app can no longer use the sync feature unless we keep the official sync endpoint built in.

**Comment by jokester** [comment]
> move all the sync actions that were previously done on the client side to the server side

Does that mean `/cors` API can be removed ? Because I don't see its other uses if browser can do sync without CORS-capable storage backends.

**Comment by fred-bf** [comment]
> fix: remove corsFetch

cors api already removed in commit https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/pull/4285/commits/eebc334e02e9f5d9f83203c97fbf4622a9141d0a

**Comment by fred-bf** [comment]
@Yidadaa for native client we can use tauri's http client to perform request without CORS issue. but it may take a while to refactor the code. My idea is that we release a patch version first, and then migrate the whole thing to the new interface.

**Comment by H0llyW00dzZ** [comment]
> Let's turn it into a RPC, here is a proposal:
> 
> ```ts
> type WebDavConfig = {
>   type: 'webdav',
>   endpoint: string,
>   username: string,
>   password: string
> }
> 
> type UpstashConfig = {
>   type: 'upstash',
>   endpoint: string,
>   username: STORAGE_KEY,
>   apiKey: string
> }
> 
> type SyncAction = {
>   type: 'check'
> } | {
>   type: 'get',
>   key: string,
> } | {
>   type: 'set',
>   key: string,
>   value: string
> }
> 
> type SyncPayload = {
>   config: WebDavConfig | UpstashConfig,
>   action: SyncAction
> }
> 
> // client side
> function createUpstashClient() {
>   return async check() {
>     return await call('/api/sync', { /* webdav config */ })
>   },
>   // get(key), set(key, value)
> }
> 
> // server side, /api/sync/route.ts
> async function handle(req) {
>   const payload = JSON.parse(await req.json()) as SyncPayload;
> 
>   // dispatch payload to server side check/get/set actions
> }
> ```
> 
> The main idea is to move all the sync actions that were previously done on the client side to the server side. The downside is that the desktop app can no longer use the sync feature unless we keep the official sync endpoint built in.

It would be better to keep the official sync that can be used in the desktop app because it is more secure and safe (essentially built-in and isolated) in the desktop environment.

**Comment by fred-bf** [comment]
Merge the current PR first, and if there is any other questions, you could comment here or file a new issue

**Pull request merged**